### ISSUE TITLE
PR Draft: Unary and inline assignment expressions

### ISSUE THREAD

**Issue by Janne252** [Perryvw/TypescriptToLua#162]
```ts
let count = 0;
count++;

let count01 = count++;
let count02 = count += 1;
let count03 = count = count + 1;

let counter01 = `Next count: ${++count}`;
let counter02 = `Next count: ${count = count + 1}`;
```
Transpiles to 
```lua
local count = 0
count=count+1

-- unexpected symbol near '='
local count01 = count=count+1
-- unexpected symbol near '='
local count02 = count = count+1
-- unexpected symbol near '='
local count03 = count = (count+1)

-- ')' expected near '='
local counter01 = "Next count: "..tostring(count=count+1)..""
-- ')' expected near '='
local counter02 = "Next count: "..tostring((count = (count+1)))..""
```
This is an issue because lue doesn't support assignment in an expression.

Proposal: Wrap the assignments in an anonymous function call when the context requires it:

```lua
local count = 0
count=count+1
-- Left becomes its own statement and operand becomes the return value
local count01 = (function() count=count+1; return count end)()
local count02 = (function() count=count+1; return count end)()
local count03 = (function() count = (count+1); return count end)()

local counter01 = "Next count: "..tostring((function() count=count+1; return count end)())..""
-- Extra parenthesis around the assignment causes and error, changed:
--     (count = (count+1)); -> count = (count+1);
local counter02 = "Next count: "..tostring((function() count = (count+1); return count end)())..""
```

**Comment by Perryvw** [comment]
We already dealt with this for returns, but I like how this method could be used in any situation since it just replaces an expression with another expression.

**Comment by Janne252** [comment]
@Perryvw Could you provide an example of that, please?

**Comment by Perryvw** [comment]
For return statements we simply transpile `return a += b` to `a = a + b; return a;` But I think that method will not work in every case. The solution you suggested I think will work in ever case since you replace an expression node with another expression node. We could use that to also rework the transpileReturn solution right now.

`TSHelper.isBinaryAssignmentToken` can be reused for this, and should probably get a similar `isUnaryAssignmentToken`

**Comment by Perryvw** [comment]
The way to do it is to probably construct a new AST node and transpile that, to ensure it also properly handles set and get accessors. It would work something like:
```
const [isAssignment, operator] = tsHelper.isBinaryAssignmentToken(operator);
if (isAssignment) {
    const expr = ts.createBinary(node.left, operator, node.right);
    const return = ts.createReturn(node.left);
    return `(function() ${expr} ${return})()`
}
```

Alternatively we could go even deeper and create a function node with the expression and return statement, then put that in a call node and transpile that, might be nicer.

**Comment by Janne252** [comment]
Good points, I'll experiment with that.

**Comment by Janne252** [comment]
#### TL;DR; Proved to be more difficult than I anticipated, releasing the "claim" to implement this ticket.

transpilePostfixUnaryExpression is a bit more complicated because we have to return the value before the operation:
**TS**
```ts
let count = 0;
count = count++;
```
**Lua**
```lua
local count = 0
count = (function() local count_before = count; count = count + 1; return count_before end)()
```

Not sure if this creates an issue because the value returned is actually a temp var declared in the anonymous function. It seems that an assignment cannot be chained after that, e.g. `count++ += 1`, which means the stored temp var is probably safe.

I looked into transpiling unaries to anonymous functions when the context requires it but didn't get results that I wanted. I got it to work but there's some duplication (multiple nested function calls) and it started to look like a hack. I don't have enough understanding of TypeScript compiler so it's probably better for someone else to implement this.

Binary expressions that contain assignments were even more difficult to get to work, e.g.
**TS**
```ts
let count = 0;
count = count += 1;
```
**Lua**
```lua
local count = 0
count = (function() count = count + 1; return count end)()
```

**Comment by Janne252** [comment]
Gave this another go, managed to refactor unary expressions to replace it with another node:
**transpilePostfixUnaryExpression**
```ts
const tempVarName = "_before";
const expression = ts.createBinary(
    node.operand,
    ts.SyntaxKind.EqualsToken,
    ts.createImmediatelyInvokedArrowFunction([
        ts.createVariableStatement(
            null,
            [ts.createVariableDeclaration(tempVarName, null, node.operand)]
        ),
        ts.createStatement(
            ts.createBinary(
                node.operand,
                ts.SyntaxKind.EqualsToken,
                ts.createBinary(node.operand, operator, ts.createLiteral(1))
            )
        ),
        ts.createReturn(ts.createIdentifier(tempVarName)),
    ])
);
```

Which beautifully transpiles `count01++;` to 

```lua
count01 = (function()
    local _before = count01

    count01 = (count01+1)
    return _before
end
)()
```

**Comment by Perryvw** [comment]
I think you were on the right path, I will have a look at it when I can. I think `count++` might be the exception by the way, I think all other ones return the value after applying the operator.

**Comment by Janne252** [comment]
The most difficult part for me is to figure out a way to only wrap the necessary part in a function call.
All the exampels below share the same pattern - assingment. Only difference is that the RHS of the expression has to be wrapped in a function call when it contains an assignment.
```ts
let count = 0;
cout++; // pure unary
count = count++; // unary in an assigment
count = count = count  +1; // chained assignmet (this can go on forever)

--count;
count = --count;
```

```lua
local count = 0
count = count + 1
count = (function() count = count + 1; return count end)()

count = count - 1
count = (function() count = count - 1; return count end)()
```

I think `isBinaryAssignmentToken` should also return `[true, ts.SyntaxKind.EqualsToken]` when it receives `EqualsToken`, since that is an assignment? Problem is this causes an infinite recursive loop as a node generated in `transpileBinaryExpression` gets detected as assignment, which creates a new one, gets detected again, etc.

**Comment by Perryvw** [comment]
Proposal

**Comment by Perryvw** [comment]
Oops missclick.

**Comment by Perryvw** [comment]
**Proposal**
Replace assigning Binary expressions (+=, -=, *=, ... etc) and assigning Unary postfix/prefix expressions (++, --) to immediately-invoked functions. This gives a neat way to transpile previously impossible statements, since you replace an expression node in the AST with another expression node, guarateeing the integrity of the translation.

Let's look at some examples:

**Current Situation**
`let nextItem = array[++i];` Translates to `local nextItem = array[i = i + 1]` which is a syntax error in Lua.

A binary example could be `a => result += a` which translates to `function(a) return result = result + a end` which again is a syntax error.

**Proposed situation**
Really `++i` is simply shorthand for `i = i + 1; return i`. This means we could change our translation to become `local nextItem = array[(function() i = i + 1; return i end)()]`, which is valid Lua.

The binary example would translate to `function(a) return (function() result = result + a; return result end)() end`.

**Alternative**
We could just not allow assignment operator expressions anywhere other than statements or simple lambda functions and throw an exception otherwise.

**Advantages**
We can transpile a larger part of the TS spec without throwing random unexpected errors.

**Downsides**
Increases the complexity of generated lua quite a bit, would need to specially deal with statements using this since transpiling a statement to `(function() end)()` will be interpreted by Lua as calling the previous statement value with a function parameter.

**Comment by Janne252** [comment]
### Unary solution A (@Perryvw's comment, the original issue description, #162)

Transpiling assignments into immediately executed functions also causes a change in order of operation if the LHS of the expression's precedence is less than function call. This is demonstrated and implemented in #164, but not really a high priority as it only seems to matter in `i + i++`, root cause of it being the return value of `i++` being the value before incrementation. 

#164 also implements a check where ++/-- unary expression is only wrapped in a function call when the outer node doesn't support assignment statements. This includes expressions like `i = j++`, `'Count: ' + ++i`, etc. Basically anything where the unary expression is part of an expression and not just alone, i.e. `i++`.

### Unary solution B
I wonder if ++/-- unary expressions could be transpiled to a chained expression, e.g. 
- `i++` -> `i = i + 1`
- `i = ++j` => `i = j = j + 1`
- `i + i++` => unary transpiler handles this on its own (doesn't pass it on as a chained assignment)

This way unaries are transpiled into a series of assignments. Then BinaryExpression transpilation would have to take care of wrapping the necessary parts in function calls (i.e. handle chained assignments).
This doesn't solve `i++`, of course. That's why it has to be transpiled separately based on context. 

The benefit of transpiling prefix unaries into chained assignmentes is that you'd effectively solve 50% of unary expressions that require special treatment & have to implement support for chained expressions.
`i++` would need its own  solution that checks if the outer scope supports assignments, just like unary solution A. 

### Conclusion
I think B is a better approach because it links closely related expressions together and reduces code duplication:
- **A)** Unary checks if context requires function wrapper AND eventually binaryExpression doing it as well when it supports chained assignments, no support for chained assignments for now 😢 
- **B)** Unary ONLY checks if context requires function wrapper when  it's a `i++` unary. binaryExpression handles all chained assignments. Unaries supported & chained assignments supported 👍

**Comment by Perryvw** [comment]
I would prefer a chained expression but those are not supported by Lua.

**Comment by Perryvw** [comment]
Addressed by merging #166
